File: 0386-lexicographical-numbers/0386-lexicographical-numbers.kt
class Solution {
    fun lexicalOrder(n: Int): List<Int> {
        val answer = mutableListOf<Int>()

        for (i in 1..9) {
            if (i <= n) {
                answer.add(i)
                this.generate(i.toString(), n, answer)
            }
        }
        return answer
    }

    private tailrec fun generate(prefix: String, limit: Int, answer: MutableList<Int>) {
        for (i in 0..9) {
            val candidate = "${prefix}${i}"
            val candidate_num = candidate.toInt()

            if (candidate_num <= limit) {
                answer.add(candidate_num)
                this.generate(candidate, limit, answer)
            }
        }
    }
}

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.py
class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        return list(self._solve(n))

    def _solve(self, n: int):
        for i in range(1, 10):
            if i <= n:
                yield i
                yield from self._generate(str(i), n)

    def _generate(self, prefix: str, limit: int):
        for i in range(0, 10):
            candidate = f"{prefix}{i}"
            candidate_num = int(candidate)
            if candidate_num <= limit:
                yield candidate_num
                yield from self._generate(candidate, limit)
        return